runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -307,6 +318,95 @@ int32_t TR_EscapeAnalysis::perform()
 #endif
 
    cost = performAnalysisOnce();
+
+   if (!_callsToProtect->empty() && manager()->numPassesCompleted() < _maxPassNumber)
+      {
+      TR::CFG *cfg = comp()->getFlowGraph();
+      TR::Block *block = NULL;
+      TR::TreeTop *lastTree = comp()->findLastTree();
+      TR_EscapeAnalysisTools tools(comp());
+      RemainingUseCountMap *remainingUseCount = new (comp()->trStackMemory()) RemainingUseCountMap(RemainingUseCountMapComparator(), comp()->trMemory()->currentStackRegion());
+      for (TR::TreeTop *tt = comp()->findLastTree(); tt != NULL; tt = tt->getPrevTreeTop())
+         {
+         TR::Node *node = tt->getNode();
+         if (node->getOpCodeValue() == TR::BBEnd)
+            block = node->getBlock();
+         else if (node->getOpCode().isCheck() || node->getOpCode().isAnchor() || node->getOpCodeValue() == TR::treetop || node->getOpCode().isStoreDirect())
+            node = node->getFirstChild();
+
+         auto nodeLookup = _callsToProtect->find(node);
+         if (nodeLookup == _callsToProtect->end()
+             || TR_EscapeAnalysisTools::isFakeEscape(node)
+             || node->getSymbol() == NULL
+             || node->getSymbol()->getResolvedMethodSymbol() == NULL
+             || node->getSymbol()->getResolvedMethodSymbol()->getResolvedMethod() == NULL)
+            continue;
+
+         if (node->getReferenceCount() > 1)
+            {
+            auto useCount = remainingUseCount->find(node);
+            if (useCount == remainingUseCount->end())
+               {
+               (*remainingUseCount)[node] = node->getReferenceCount() - 1;
+               continue;
+               }
+            if (useCount->second > 1)
+               {
+               useCount->second -= 1;
+               continue;
+               }
+            }
+
+         if (!performTransformation(comp(), "%sHCR CALL PEEKING: Protecting call [%p] n%dn with an HCR guard and escape helper\n", OPT_DETAILS, node, node->getGlobalIndex()))
+            continue;
+
+         ((TR_EscapeAnalysis::PersistentData*)manager()->getOptData())->_processedCalls->set(node->getGlobalIndex());
+
+         _repeatAnalysis = true;
+
+         optimizer()->setValueNumberInfo(NULL);
+         cfg->invalidateStructure();
+
+         TR::TreeTop *prevTT = tt->getPrevTreeTop();
+         TR::Block *callBlock = block->split(tt, comp()->getFlowGraph(), true, true);
+
+         // check uncommoned nodes for stores of the candidate - if so we need to add the new temp to the
+         // list of loads
+         for (TR::TreeTop *itr = block->getExit(); itr != prevTT; itr = itr->getPrevTreeTop())
+            {
+            if (itr->getNode()->getOpCodeValue() == TR::astore
```

Comment:
A store can be under a treetop node too in theory. Perhaps calling the getStoreNode() api is a better option here ?

Reply:
Addressed in update